### PR TITLE
[ot_certs] Add support for substitutions in a template

### PR DESCRIPTION
**Pull request by pamaury**
This PR does two things (details below):
- introduce a way to substitute values into a template (and output a new template), those values can be stored in a json file
- refactor the hjson parsing code so that the same code is used for the template and substitution values: this clarifies and simplifies the way the values are parsing (the current code is quite hard to follow)

Test with:
```bash
# generate another template
./bazelisk.sh run //sw/host/opentitantool -- certificate subst $PWD/sw/device/silicon_creator/lib/cert/uds.hjson $PWD/sw/device/silicon_creator/lib/cert/uds_example_data.json
# substitute and generate DER
./bazelisk.sh run //sw/host/opentitantool -- certificate generate $PWD/sw/device/silicon_creator/lib/cert/uds.hjson --subst $PWD/sw/device/silicon_creator/lib/cert/uds_example_data.json $PWD/example.pem
openssl x509 -inform DER -in $PWD/example.pem -noout -text
```

**Details:**
Presently, the hjson template parser does too many things at once: for the field it both parses the raw hjson value AND converts those values to whatever type the template field as (e.g. the `layer` is an integer) in a single step. This is achieved by some nontrivial serde code that calls into the `hjson` module. As a result, it's actually quite hard to understand what formats are accepted for a field or to write a unittest. This PR separates the parsing into two steps:
- First, the (h)json value is parsed without any interpretation as a string, integer or array of integer (`SubstValue`): this means that literally only syntax errors or obviously wrong structures will lead to a parsing error.
- Then, this raw value is converted to the actual type of the field using the `ConvertValue` trait: this means that all the logic about what formats are accepted and how to interpret them is entirely contained in that trait and easily testable or reusable.

### CODE REVIEW

**Review by jwnrt — APPROVED** [review]
LGTM! Thanks @pamaury.

We discussed the `// TODO` and agreed that it wasn't important enough to solve as we won't run into its edge case for our templates. It won't cause errors, just conservative failures.

**Review by timothytrippel — APPROVED** [review]

**Review by jwnrt — APPROVED** [review]
Updates LGTM, serialization is simpler and the documentation is nice.

**Comment by pamaury** [comment]
The CI failure looks unrelated (download failure).

**Comment by pamaury** [comment]
I have added some unittest (and caught a couple of bugs in passing).
